### ISSUE TITLE
hierarchical package presentation in Ceylon Explorer

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-ide-eclipse#1238]
The Ceylon Explorer does not currently support `Package Presentation > Hierarchical`. I quickly tried to make it work but failed. One for @davidfestal.

**Comment by gavinking** [comment]
Related #1210.

**Comment by lukedegruchy** [comment]
I've had my package presentation set to hierarchical (it's my personal preference) for months and I've had no issues.   Granted my Ceylon projects are relatively simple (no Maven, OSGi or Java).

**Comment by gavinking** [comment]
Yeah, um, that's weird. AFAICT, it seems to be working great for me now too. OK, good.

Perhaps I was thinking that modules should be displayed hierarchically just like packages? i.e. that all the SDK modules would appear beneath a `ceylon` node?

@lukedegruchy as a user of hierarchical presentation do you think that would be better or worse than the current behavior which is to display only packages hierarchically?

**Comment by lukedegruchy** [comment]
I've thought about it for a bit and I decided that hierarchical module presentation would look better than flat.   Looking at the ceylon-sdk project is what made up my mind.   The ceylon.**** modules didn't clearly break from the com.redhat.ceylon.*** modules, which is what I'd prefer given my mental model.  Also, it's probably better for the sake of consistency with how packages are currently displayed.

**Comment by gavinking** [comment]
So let's leave this issue open then.

**Comment by lukedegruchy** [comment]
Actually, one thing I've just noticed (didn't notice this for months even though it was staring me in the face) is that while hierarchical packages do work, they don't work entirely as expected.

A Java project with hierarchical package display shows the parent package and child packages without the parent package.

A Java project would display packages this way:

> parent
>> child1
>> child2
>> child3

In the Ceylon IDE, it displays like this:

> parent
>> parent.child1
>> parent.child2
>> parent.child3

So ideally, Ceylon packages in a hierarchical view should be displayed as in they are in a Java project (the first example in my comment).

This is the case in my Ceylon 1.1 IDE, so nothing has changed.

**Comment by gavinking** [comment]
Ah, well I guess _that_ should be pretty trivial to fix.

**Comment by gavinking** [comment]
@lukedegruchy do you prefer it like that?

**Comment by lukedegruchy** [comment]
Yes, I do.

**Comment by lukedegruchy** [comment]
I mean, I prefer:

> parent
>> child1
>> child2
>> child3

**Comment by lukedegruchy** [comment]
I see you already fixed the parent/child package presentation issue.  Nice!  :)

**Comment by lukedegruchy** [comment]
There's one thing that I just noticed.

> path.to.parent (module)
>> child1 (package)
>>> grandchild1  (package)

>>child2  (package)
>>> grandchild2  (package)

As opposed to:

> path.to.parent (module)
>> path.to.parent.child1  (package)

>>> grandchild1  (package)

>> path.to.parent.child2  (package)
>>> grandchild2  (package)

I have a slight preference for the latter (fully qualified parent package) over the former, but I don't consider it to be a showstopper by any means.

**Comment by jvasileff** [comment]
:+1: *very* nice!

One thing I noticed is that now in the Ceylon perspective `src` folders appear at the end of the list (alphabetical order, I guess). Switching to the Java perspective promotes/special cases source folders and puts them at the top, which is better.

**Comment by jvasileff** [comment]
Here's something that's truly awful (but also a JDT bug). If the tree has focus, and you press a key, say `r`, to jump to a file like, `run.ceylon`, it works. *Unless* `run.ceylon` has `>` prefix indicating uncommitted changes. Then, `r` just does something crazy and you have to press the `>` key to navigate to the file.

**Comment by davidfestal** [comment]
@jvasileff : this is a bug related to the way team-related text decorators are added to the node label in the Common Navigator Framework. This was also the root cause of #1454. However in this case I'm not sure it's fixable.

**Comment by jvasileff** [comment]
Thanks @davidfestal. I'll just give "Team -> Disconnect" a try then, until the Eclipse folks can sort it out :smile:

**Comment by jvasileff** [comment]
> src folders appear at the end of the list 

Well, after going to "Customize View" -> "Content" and toggled Resources off and on, the `src` folder has floated back up to the top?!

**Comment by davidfestal** [comment]
@jvasileff : I'm looking into this right now. However could you open another issue for this ?
